### PR TITLE
Spike:  load json events into Vertica

### PR DESCRIPTION
**Pull request by brianhw**
Here is a spike implementing Vertica loading for events as specified in DE-17.  

* Note that the raw event is written out to a string of length 60000, rather arbitrarily chosen.  There is no logic included to write to a binary blob on Vertica.  
* No special work has been done to determine when course_id information is stored in the label field of segment events.  
* The "date" field was left in from before, even though it wasn't in the spec.  
* Raw events are dumped using json.dumps(event, sort_keys=True), instead of cjson.  Part of the reason was simply to make testing easier, by making the output more canonical.
* Pruning task was left in, because it also grants access.   The pruning was disabled in code.   
* Validation for datetime strings was added, to deal more robustly with bogus date values without rejecting the entire record.

### CODE REVIEW

**Review by tobz — APPROVED** [review]
Seems pretty straightforward to me.

:shipit: 👍

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 1]
```diff
@@ -707,10 +795,14 @@ def mapper(self, line):
             return
 
         # Ignore events that begin with a slash (i.e. implicit events).
+        event_url = None
         if event_type.startswith('/'):
-            self.incr_counter(self.counter_category_name, 'Discard Implicit Events', 1)
-            return
+            # self.incr_counter(self.counter_category_name, 'Discard Implicit Events', 1)
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:819]: You can probably get rid of this commented piece.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 2]
```diff
@@ -731,32 +823,45 @@ def mapper(self, line):
             self.incr_counter(self.counter_category_name, 'Discard Missing Event Source', 1)
             return
 
-        if (event_source, event_type) in self.known_events:
-            event_category = self.known_events[(event_source, event_type)]
-        else:
-            event_category = 'unknown'
+        # if (event_source, event_type) in self.known_events:
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:850]: And nix these comments, too.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 3]
```diff
@@ -731,32 +823,45 @@ def mapper(self, line):
...
+        #     event_category = self.known_events[(event_source, event_type)]
+        # else:
+        #     event_category = 'unknown'
 
         project_name = self.PROJECT_NAME
 
-        event_dict = {'version': VERSION}
+        # event_dict = {'version': VERSION}
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:857]: Nix this, too, probably.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 4]
```diff
@@ -731,32 +823,45 @@ def mapper(self, line):
...
-        self.add_calculated_event_entry(event_dict, 'project', project_name)
+        # was project
+        self.add_calculated_event_entry(event_dict, 'source', project_name)
         self.add_calculated_event_entry(event_dict, 'event_type', event_type)
-        self.add_calculated_event_entry(event_dict, 'event_source', event_source)
-        self.add_calculated_event_entry(event_dict, 'event_category', event_category)
+
+        # was event_source
+        self.add_calculated_event_entry(event_dict, 'emitter_type', event_source)
+        # self.add_calculated_event_entry(event_dict, 'event_category', event_category)
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:867]: ...and this.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 5]
```diff
@@ -731,32 +823,45 @@ def mapper(self, line):
...
+        self.add_calculated_event_entry(event_dict, 'course_id', course_id)
         self.add_calculated_event_entry(event_dict, 'username', username)
 
-        self.add_agent_info(event_dict, event.get('agent'))
+        self.add_calculated_event_entry(event_dict, 'raw_event', json.dumps(event, sort_keys=True))
+
+        if event_url is not None:
+            self.add_calculated_event_entry(event_dict, 'url', event_url)
+
+        # self.add_agent_info(event_dict, event.get('agent'))
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py]: nix

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 6]
```diff
@@ -1003,22 +1115,28 @@ def mapper(self, line):
 
         self.incr_counter(self.counter_category_name, u'Subset Project {}'.format(project_name), 1)
 
-        event_dict = {'version': VERSION}
+        # event_dict = {'version': VERSION}
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:1180]: more commented lines to remove over this next chunk of code.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 7]
```diff
@@ -1272,7 +1390,8 @@ def queries(self):
             ),
         ]
         # Check for pruning.
-        if self.interval and self.earliest_date and self.retention_interval:
+        # if self.interval and self.earliest_date and self.retention_interval:
+        if False:
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py:1507]: :) Do...something else here.

File: edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py [hunk 8]
```diff
@@ -1,14 +1,18 @@
...
+import ciso8601
 from ddt import ddt, data
 import luigi
 
 from edx.analytics.tasks.common.tests.map_reduce_mixins import MapperTestMixin
+from edx.analytics.tasks.util import eventlog
 from edx.analytics.tasks.util.tests.opaque_key_mixins import InitializeOpaqueKeysMixin
 from edx.analytics.tasks.warehouse.load_internal_reporting_events import (
-    EventRecord,
+    # EventRecord,
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py:14]: nix

File: edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py [hunk 9]
```diff
@@ -87,39 +91,71 @@ def create_task(self, date=None):  # pylint: disable=arguments-differ
...
+        self.assertEquals(url, kwargs['event_type'])
+        self.assertEquals(getattr(actual_record, 'event_type'), 'edx.server.request')
+
     def test_problem_check(self):
         template = self.event_templates['problem_check']
         event = self.create_event_log_line(template=template)
         expected_key = (self.DEFAULT_DATE, self.task.PROJECT_NAME)
         expected_dict = {
-            'version': VERSION,
+            # 'version': VERSION,
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py:134]: more old, commented lines to get rid of

File: edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py [hunk 10]
```diff
@@ -313,39 +360,42 @@ def test_android_screen(self):
         event = self.create_event_log_line(template=template)
         expected_key = (self.DEFAULT_DATE, self.DEFAULT_PROJECT)
         expected_dict = {
-            'version': VERSION,
+            # 'version': VERSION,
```
Comment (iloveagent57) [inline edx/analytics/tasks/warehouse/tests/test_load_internal_reporting_events.py:383]: you guessed it, more comments to axe

**Review by iloveagent57 — COMMENTED** [review]
You'll notice a theme with my suggestions.  Looks good, otherwise.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py [hunk 11]
```diff
@@ -1003,22 +1150,60 @@ def mapper(self, line):
...
 
-        record = EventRecord(**event_dict)
+        # Try to extract course_id information.
+        course_id = event_dict.get('course_id')
+        if course_id is None:
+            # course_id may be stored in 'label', so try to parse what is there.
+            label = event_dict.get('label')
+            if is_valid_course_id(label):
+                self.add_calculated_event_entry(event_dict, 'course_id', label)
+                course_id = event_dict.get('course_id')
```
Comment (mulby) [inline edx/analytics/tasks/warehouse/load_internal_reporting_events.py]: I'm confused about this - did you mean to set `course_id = label`?

**Review by mulby — COMMENTED** [review]
Given that we aren't planning on merging this yet - LGTM! Great spike - if/when we decide to actually go this route we can come back in and clean this stuff up.

File: edx/analytics/tasks/warehouse/load_internal_reporting_events.py (hunk 11 above)
Comment (brianhw) [reply edx/analytics/tasks/warehouse/load_internal_reporting_events.py]: Well, it does the right thing, but may not be the clearest.   The previous line puts label into event_dict as 'course_id', performing whatever validation is appropriate.   So just to be safe here, I set the local value of course_id to the value that's actually in the event_dict.  It should be the same thing.....

**Comment by brianhw** [comment]
This has been superceded by PR #436.

**Pull request closed without merging**